docs: add docs for configuring and using plugin (#57)
cultureamp/migrations-runner-buildkite-plugin: README.md

# Migrations Runner Plugin

> This plugin is designed to run database migrations task, without requiring consumer to write and maintain their own scripts.

Reviewer: ```suggestion
This plugin is designed to run database migrations task, without requiring consumers to write and maintain their own scripts.
```

- A Parameter Store parameter extending the task definition by providing entrypoint overrides and networking configuration
- A log group for the task
- A security group for your service (this can be the [base-infrastructure-for-services](https://github.com/cultureamp/base-infrastructure-for-services) source security group)

This can be visualised below:

![The overall flow of this plugin and AWS resources](docs/images/diagram.svg)

## Example

Add the following lines to your `pipeline.yml`:

```yml
steps:
  - label: "Run my very cool migration task"
    plugins:
      - cultureamp/aws-assume-role:
        role: arn:aws:iam::${AWS_ACCOUNT_ID}:role/deploy-role-cool-service
        region: ${REGION}
      - cultureamp/migrations-runner#v1.0.0:
          parameter-name: "/cool-service/cool-farm/migrations-runner-config"
          command: "/bin/migrate"
```

## Configuration

### `parameter-name` (Required, string)

The name of the parameter in Parameter Store that contains the task definition. This will be setup by the `MigrationsRunner` construct, so refer to that to find your specific parameter name.

### `command` (Optional, string)

> The name of the command or script to run in the task. When omitted, the task will run the command specified in the container's `CMD`.

Reviewer: To clarify: There's also the case of using the `ENTRYPOINT` directive. If that's used, the `command` parameter appends to the ENTRYPOINT command as a kind of override. I don't think it works when `CMD` is defined, but ECS tries to add to it.

I know we threw around the idea that "command" is a kind of override. So product engineers can treat their migrations runner as an app (ENTRYPOINT) or fire and forget runner (CMD).  Is that worth digging more into later?

Author: I think what you're described here is correct, I'll add an explanation of this behaviour to the docs, as this'll change depending on how the Dockefile is written

> [!TIP]
> When specifying a command to run, ensure that the command is:
>
> - Enclosed in quotes
> - Space-separated
>
> For example if you'd like to run the following command for your migration task: `parameter-store-exec bundle exec bin/run_mongodb_migrations`
>
> You'd configure the plugin like this:
>
>```yml
>steps:
>  - plugins:
>      - cultureamp/migrations-runner#v1.0.0:
>          parameter-name: "/cool-service/cool-farm/migrations-runner-config"
>          command: "parameter-store-exec bundle exec bin/run_mongodb_migrations"
>```

### `timeout` (Optional, integer)

The timeout in seconds that the plugin will wait for the task to complete. If the task does not complete within this time, the plugin will fail. The task execution will continue to run in the background.

> [!NOTE]
> This only affects how the migration task's status is reported in the Buildkite step, and the migration task may still run successfully even if the plugin timeouts. To avoid false negatives on migration task failures, we recommend using the default value.

> 

Reviewer: 👍

Default: 2700

## Context

This plugin is based on an existing pattern in `murmur` where database migrations are run as a task on ECS. To provide additional context for how this plugin is expected to be used, this is the expected pattern:

- A CI image is built and pushed to ECR

> - The command of the image is overridden in the task definition to run the specific migration task

Reviewer: Is that true?

My understanding is the `command` in the task definition is overwritten by the parameter. The task definition comes with no `command`.

The parameter however doesn't override the `CMD` or `ENTRYPOINT` directives in the Dockerfile. It extends the latter as a series of arguments.

Author: My understanding of this could be wrong, I'll double check this now as getting this right will avoid a lot of confusion!

Author: Based off [this doc](https://pkg.go.dev/github.com/aws/aws-sdk-go-v2/service/ecs@v1.49.0/types#ContainerOverride), only the `CMD` is overridden by our plugin. I think this means that whether we override or append to the current command to execute depends on whether the image uses `CMD` or `ENTRYPOINT`, respectively. `roles-service` uses `ENTRYPOINT`, so I'll test this to verify if we append in this scenario

Author: Confirmed [here](https://buildkite.com/culture-amp/roles-service/builds/5706) that the `ENTRYPOINT` is appended to, but `CMD` is overridden [here](https://buildkite.com/culture-amp/roles-service/builds/5707). I'll add a note to the docs to distinguish between the behaviours, depending on how consumers write their Dockerfiles 🙌

# Migrations Runner Plugin

> This plugin is designed to run database migrations task, without requiring consumers to write and maintain their own scripts with executing and tracking compute services to facilitate migration.

Reviewer 2: Some grammar + this feels like it could do with a reword.

```suggestion
This plugin is designed to run database migrations tasks, without requiring consumers to write and maintain their own scripts including execution and tracking compute services to facilitate migration.
```

This plugin comes with some assumed infrastructure that needs to be deployed before it can be run. This infrastructure is as follows:

- An ECS cluster
- An ECS task definition
- An ECR image
- An IAM role for the ECS task
- An IAM role for the BK agent to start the task
- A Parameter Store parameter extending the task definition by providing entrypoint overrides and networking configuration
- A log group for the task
- A security group for your service (this can be the [base-infrastructure-for-services](https://github.com/cultureamp/base-infrastructure-for-services) source security group)

This can be visualised below:

![The overall flow of this plugin and AWS resources](docs/images/diagram.svg)

## Example

Add the following lines to your `pipeline.yml`:

```yml
steps:
  - label: "Run my very cool migration task"
    plugins:
      - cultureamp/aws-assume-role:
        role: arn:aws:iam::${AWS_ACCOUNT_ID}:role/deploy-role-cool-service
        region: ${REGION}
      - cultureamp/migrations-runner#v1.0.0:
          parameter-name: "/cool-service/cool-farm/migrations-runner-config"
          command: "/bin/migrate"
```

## Configuration

### `parameter-name` (Required, string)

> The name of the parameter in Parameter Store that contains the task definition. This will be setup by the `MigrationsRunner` construct, so refer to that to find your specific parameter name.

Reviewer 2: Is it worth providing some limited commentary about this?

I am thinking specifically that it will always end in `migrations-runner-config` if you are using the `MigrationsRunner` construct.

Author: Good idea, will add some guidance describing this

### `command` (Optional, string)

> The name of the command or script to run in the task. When omitted, the task will run the command specified in the container's `CMD`.

Reviewer 2: ```suggestion
The name of the command or script to run in the task. When omitted, the task will run the command specified in the container's `CMD` or `ENTRYPOINT`.
```

Small suggestion, just to clarify that you can still use ENTRYPOINT with this, regardless of whether or not you specify `command`.

Author: Good callout, I think the details below explain this is more detail!

> [!NOTE]
> Care must be taken, as depending on whether the Dockerfile used to create the referenced image has a `CMD` or `ENTRYPOINT` instruction, the command to execute in the container will either be overridden or appended to, respectively.
>
> To illustrate these behaviours, consider the following scenarios:
>
> - A Dockerfile with an `ENTRYPOINT` of `./run-binary`. Leaving the plugin's command argument blank will simply execute `./run-binary` in the container. Inputting `run-sub-command` in the plugin's command argument will execute `./run-binary run-sub-command`.

> > - A Dockerfile with a `CMD` of `./run-binary`. Leaving the plugin's command argument blank will simply execute `./run-binary` in the container. Inputting `run-sub-command` in the plugin's command argument will execute `run-sub-command`.

Reviewer 2: 👍
